GameServer/Logic/Maps/SummonersRift.cs

```diff
@@ -349,36 +400,79 @@ public override bool Spawn()
                 MinionSpawnPosition.SPAWN_RED_BOT,
                 MinionSpawnPosition.SPAWN_RED_MID,
             };
-
-            if (_waveNumber < 3)
+            var cannonMinionCap = 2;
+            foreach (var pos in positions)
             {
-                for (var i = 0; i < positions.Count; ++i)
+                var _spawnToWaypoints = new Dictionary<MinionSpawnPosition, Tuple<List<Vector2>, uint>>
+                {
+                    {MinionSpawnPosition.SPAWN_BLUE_BOT, Tuple.Create(BLUE_BOT_WAYPOINTS, 0xff26ac0f)},
+                    {MinionSpawnPosition.SPAWN_BLUE_MID, Tuple.Create(BLUE_MID_WAYPOINTS, 0xffff8f1f)},
+                    {MinionSpawnPosition.SPAWN_BLUE_TOP, Tuple.Create(BLUE_TOP_WAYPOINTS, 0xff6793d0)},
+                    {MinionSpawnPosition.SPAWN_RED_BOT, Tuple.Create(RED_BOT_WAYPOINTS, 0xff9303e1)},
+                    {MinionSpawnPosition.SPAWN_RED_MID, Tuple.Create(RED_MID_WAYPOINTS, 0xff4a20f1)},
+                    {MinionSpawnPosition.SPAWN_RED_TOP, Tuple.Create(RED_TOP_WAYPOINTS, 0xffd23c3e)}
+                };
+                var waypoints = _spawnToWaypoints[pos].Item1;
+                var inhibitorId = _spawnToWaypoints[pos].Item2;
+                var inhibitor = GetInhibitorById(inhibitorId);
+                var isInhibitorDead = inhibitor.getState() == InhibitorState.Dead && !inhibitor.GetRespawnAnnounced();
+
+                var oppositeTeam = TeamId.TEAM_BLUE;
+                if (inhibitor.getTeam() == TeamId.TEAM_PURPLE)
                 {
-                    Minion m = new Minion(MinionSpawnType.MINION_TYPE_MELEE, positions[i], _laneWaypoints[i]);
-                    AddObject(m);
+                    oppositeTeam = TeamId.TEAM_PURPLE;
                 }
-                return false;
-            }
 
-            if (_waveNumber == 3)
-            {
-                for (var i = 0; i < positions.Count; ++i)
+                var areAllInhibitorsDead = AllInhibitorsDestroyedFromTeam(oppositeTeam) && !inhibitor.GetRespawnAnnounced();
+
+                var cannonMinionTimestamps = new List<Tuple<long, int>>
```

Comment:
This list should be initialized outside the for loop to avoid re-creating it every time.